### PR TITLE
CMake: MINGW implies WIN32

### PR DESCRIPTION
**Pull request by xantares**
hdf5 requires cmake 3, so we can simplify these

### CODE REVIEW

**Review by byrnHDF — COMMENTED** [review]
I'm not sure this assumption is valid, I will need to investigate as there are two use cases for mingw, 1. On windows compiling as linux and 2. On linux compiling for windows (on windows for windows would be included).

**Comment by byrnHDF** [comment]
> I'm not sure this assumption is valid, I will need to investigate as there are two use cases for mingw, 1. On windows compiling as linux and 2. On linux compiling for windows (on windows for windows would be included).

Of course I am not sure about (1) as I have not tried that environment, and might be confusing it with cygwin.

**Comment by byrnHDF** [comment]
I was confusing cygwin and mingw, however:
Platform Variables
CMake sets certain variables to true depending on the current platform and toolchain in use. These always describe the target platform.  In older versions of CMake, these were the only way of detecting the current platform.


UNIX : is TRUE on all UNIX-like OS's, including Apple OS X and
CygWin


WIN32 : is TRUE on Windows. Prior to 2.8.4 this included CygWin


APPLE : is TRUE on Apple systems. Note this does not imply the
system is Mac OS X, only that APPLE is #defined in C/C++
header files.

MINGW : is TRUE when using the MinGW compiler in Windows

MSYS : is TRUE when using the MSYS developer environment in Windows

CYGWIN : is TRUE on Windows when using the CygWin version of cmake


NOTE that there is not a mention of MINGW on linux, however, if you are using mingw on linux then WIN32 will not be set and that is the reason for the :if (WIN32 OR MINGW) uses.

**Comment by xantares** [comment]
no, MINGW is also set when crosscompiling from linux from my experience
in the doc it is said to be set as soon as the toolchain is mingw:
https://cmake.org/cmake/help/v3.18/variable/MINGW.html?highlight=mingw

**Comment by byrnHDF** [comment]
> no, MINGW is also set when crosscompiling from linux from my experience
> in the doc it is said to be set as soon as the toolchain is mingw:
> https://cmake.org/cmake/help/v3.18/variable/MINGW.html?highlight=mingw

Right, but on linux it will not set "WIN32" so you need to check for "MINGW" as well.

**Comment by xantares** [comment]
no, as the target platform is also windows, WIN32 is also set:
https://cmake.org/cmake/help/latest/variable/WIN32.html

**Comment by xantares** [comment]
maybe you're confusing host platform and target platform ?

**Comment by byrnHDF** [comment]
> maybe you're confusing host platform and target platform ?

maybe I am. Is it possible to have WIN32 not defined, yet have MINGW defined. If not then I agree with these changes. We will need to test.

**Comment by xantares** [comment]
afaik the mingw toolchain can only target the windows platform so WIN32 would always be set

**Comment by byrnHDF** [comment]
> afaik the mingw toolchain can only target the windows platform so WIN32 would always be set

okay, "if(WIN32 OR MINGW") == "if(WIN32)" - but now the tricky part - we need to differentiate between what is target settings vs build time settings - when building on linux, paths will need to be unix style (runtest.cmake). In this case we will need 
"if(WIN32 AND NOT MINGW)" checks.

**Comment by byrnHDF** [comment]
> > afaik the mingw toolchain can only target the windows platform so WIN32 would always be set
> 
> okay, "if(WIN32 OR MINGW") == "if(WIN32)" - but now the tricky part - we need to differentiate between what is target settings vs build time settings - when building on linux, paths will need to be unix style (runtest.cmake). In this case we will need
> "if(WIN32 AND NOT MINGW)" checks.

Wait that's not right either, "if(MINGW AND NOT {HOST is WINDOWS})"

**Comment by xantares** [comment]
In general there's no need to tweak paths because cmake uses the unix convention internally, except maybe when calling native tools.
I did run the tests and only a few failed, so I guess we're fine here.

**Comment by byrnHDF** [comment]
> In general there's no need to tweak paths because cmake uses the unix convention internally, except maybe when calling native tools.
> I did run the tests and only a few failed, so I guess we're fine here.

That's good.
BTW, the variable is "CMAKE_HOST_SYSTEM_NAME".

I may need to setup a MINGW on Windows test machine - I haven't had much luck previously.

File: CMakeInstallation.cmake [hunk 1]
```diff
@@ -290,7 +290,7 @@ if (NOT HDF5_EXTERNALLY_CONFIGURED AND NOT HDF5_NO_PACKAGES)
   set (CPACK_PACKAGE_ICON "${HDF_RESOURCES_EXT_DIR}/hdf.bmp")
 
   set (CPACK_GENERATOR "TGZ")
-  if (WIN32 OR MINGW)
+  if (WIN32)
```
Comment (byrnHDF) [inline CMakeInstallation.cmake:293]: I think the "ZIP" packager is available on linux - so no issue here.

File: CMakeLists.txt [hunk 2]
```diff
@@ -573,7 +573,7 @@ endif ()
 # and we need to call the executable from inside the CMake configuration
 #-----------------------------------------------------------------------------
 set (EXE_EXT "")
-if (WIN32 OR MINGW)
+if (WIN32)
```
Comment (byrnHDF) [inline CMakeLists.txt:575]: This needs checked - "if(MSVC)" might be preferable.

File: config/cmake/jrunTest.cmake [hunk 3]
```diff
@@ -49,7 +49,7 @@ endif ()
 
 message (STATUS "COMMAND: ${TEST_TESTER} -Xmx1024M -Dorg.slf4j.simpleLogger.defaultLog=${LOG_LEVEL} -Djava.library.path=\"${TEST_LIBRARY_DIRECTORY}\" -cp \"${TEST_CLASSPATH}\" ${TEST_ARGS} ${TEST_PROGRAM} ${ARGN}")
 
-if (WIN32 OR MINGW)
+if (WIN32)
```
Comment (byrnHDF) [inline config/cmake/jrunTest.cmake:52]: Another check needed - if this is on a linux host is the ENV{PATH} valid?

File: config/cmake/jrunTest.cmake [hunk 4]
```diff
@@ -131,7 +131,7 @@ if (NOT TEST_SKIP_COMPARE)
     file (READ ${TEST_FOLDER}/${TEST_REFERENCE} TEST_STREAM)
     list (LENGTH TEST_STREAM test_len)
     if (test_len GREATER 0)
-      if (WIN32 OR MINGW)
+      if (WIN32)
```
Comment (byrnHDF) [inline config/cmake/jrunTest.cmake:134]: This and the next are workarounds for tests on a windows host and might not be valid for a linux host.

File: config/cmake/jvolTest.cmake [hunk 5]
```diff
@@ -54,7 +54,7 @@ message (STATUS "USING ${TEST_VOL} ON COMMAND: ${TEST_TESTER} -Xmx1024M -Dorg.sl
 
 set (ENV{HDF5_VOL_CONNECTOR} "${TEST_VOL}")
 
-if (WIN32 OR MINGW)
+if (WIN32)
```
Comment (byrnHDF) [inline config/cmake/jvolTest.cmake:57]: same comments as jrunTest.cmake

File: config/cmake_ext_mod/grepTest.cmake [hunk 6]
```diff
@@ -46,7 +46,7 @@ endif ()
 message (STATUS "COMMAND: ${TEST_EMULATOR} ${TEST_PROGRAM} ${TEST_ARGS}")
 
 if (TEST_LIBRARY_DIRECTORY)
-  if (WIN32 OR MINGW)
+  if (WIN32)
```
Comment (byrnHDF) [inline config/cmake_ext_mod/grepTest.cmake:49]: Same comments as jrunTest.cmake

File: config/cmake_ext_mod/runTest.cmake [hunk 7]
```diff
@@ -38,7 +38,7 @@ endif ()
 message (STATUS "COMMAND: ${TEST_EMULATOR} ${TEST_PROGRAM} ${TEST_ARGS}")
 
 if (TEST_LIBRARY_DIRECTORY)
-  if (WIN32 OR MINGW)
+  if (WIN32)
```
Comment (byrnHDF) [inline config/cmake_ext_mod/runTest.cmake:41]: Same comments as jrunTest.cmake

**Review by byrnHDF — COMMENTED** [review]
Looks like java and plugin tests will need investigation because of use of path and classpath changes.
Otherwise everything looks good.

File: CMakeLists.txt (hunk 2 above)
Comment (xantares) [reply CMakeLists.txt:575]: yes, also EXE_EXT can be dropped

File: config/cmake/jrunTest.cmake (hunk 3 above)
Comment (xantares) [reply config/cmake/jrunTest.cmake:52]: yes, it will be used by wine

File: config/cmake/jrunTest.cmake (hunk 4 above)
Comment (xantares) [reply config/cmake/jrunTest.cmake:134]: if the issue is crlf, you will get the with wine, so this is fine

File: config/cmake_ext_mod/runTest.cmake (hunk 7 above)
Comment (xantares) [reply config/cmake_ext_mod/runTest.cmake:41]: ok

**Review by byrnHDF — APPROVED** [review]
These changes look good, if at some time in the future this gets tested on a windows machine with mingw and issues are found - a new PR can be discussed.

**Review by lrknox — APPROVED** [review]

**Review by derobins — APPROVED** [review]

**Pull request merged**